### PR TITLE
UPLOAD: Support multipart uploads and files > 5GB

### PR DESCRIPTION
**Pull request by alexpaulzor**

### CODE REVIEW

File: client/test/test_archive.py [hunk 1]
```diff
@@ -29,6 +29,21 @@ def test_push_file(archive, random_metadata, tmpfile, s3_key):
     assert common_keys == set(random_metadata.keys())
 
 
+def test_push_large_file(
+        monkeypatch, archive, random_metadata, tmpfile, s3_key):
+    monkeypatch.setenv('DATALAKE_CHUNK_SIZE_MB', 1)
+    expected_content = ('mwahaha' * 1024 * 1024).encode('utf-8')
```
Comment (bcavagnolo) [inline client/test/test_archive.py:35]: Can we randomize the input more and ensure that it's not aligned to a chunk boundary? Chunk-aligned data can obfuscate bugs.

File: client/test/test_archive.py [hunk 2]
```diff
@@ -29,6 +29,21 @@ def test_push_file(archive, random_metadata, tmpfile, s3_key):
...
+    f = tmpfile(expected_content)
+    url = archive.prepare_metadata_and_push(f, **random_metadata)
+    from_s3 = s3_key(url)
+    assert from_s3.get_contents_as_string() == expected_content
+    metadata = from_s3.get_metadata('datalake')
+    assert metadata is not None
+    metadata = json.loads(metadata)
+    common_keys = set(metadata.keys()).intersection(random_metadata.keys())
+    assert common_keys == set(random_metadata.keys())
```
Comment (bcavagnolo) [inline client/test/test_archive.py:44]: Can you check the metadata values too?

File: client/datalake/archive.py [hunk 3]
```diff
@@ -170,8 +177,26 @@ def push(self, f):
...
+            while chunk_leftover_b <= 0:
+                part = mp.upload_part_from_file(f, chunk, size=CHUNK_SIZE())
+                chunk_leftover_b = CHUNK_SIZE() - part.size
+                chunk += 1
+        except:  # NOQA
+            # Any exception we want to attempt to cancel_upload, otherwise
+            # AWS will bill us every month indefnitely for storing the
+            # partial-uploaded chunks.
+            mp.cancel_upload()
+            raise
```
Comment (bcavagnolo) [inline client/datalake/archive.py:197]: Nice.

File: client/datalake/archive.py [hunk 4]
```diff
@@ -170,8 +177,26 @@ def push(self, f):
 
     def _upload_file(self, f):
         key = self._s3_key_from_metadata(f)
-        key.set_metadata(METADATA_NAME, json.dumps(f.metadata))
-        key.set_contents_from_file(f)
+        # key.set_metadata(METADATA_NAME, json.dumps(f.metadata))
+        # key.set_contents_from_file(f)
```
Comment (bcavagnolo) [inline client/datalake/archive.py:181]: dead code?

File: client/test/conftest.py [hunk 5]
```diff
@@ -16,7 +16,7 @@
 from moto import mock_s3
 import boto
 from six.moves.urllib.parse import urlparse
-from datalake.tests import random_metadata, tmpfile  # noqa
+from datalake.tests import random_metadata, random_word, tmpfile  # noqa
```
Comment (alexpaulzor) [inline client/test/conftest.py:19]: Whoops, will revert this before merging.

Comment (bcavagnolo) [inline client/test/conftest.py:19]: Looks like you imported `random_word` but didn't use it to generate the data?

**Comment by dtkav** [comment]
Hi friends!

**Pull request merged**